client/api/src/backend.rs

```diff
@@ -418,15 +418,28 @@ pub trait Backend<Block: BlockT>: AuxStore + Send + Sync {
 	) -> sp_blockchain::Result<()>;
 
 	/// Commit block insertion.
-	fn commit_operation(&self, transaction: Self::BlockImportOperation) -> sp_blockchain::Result<()>;
+	///
+	/// The `finalization_check` closure checks that if an imported block is able to be
+	/// finalized. If the check returns false, the import succeeds, but with the imported block
+	/// unfinalized.
+	fn commit_operation<F: Fn(&Block::Hash) -> bool>(
```

Comment:
Why not pass a flag instead of a closure?

Reply:
This is for finalization check -- when `Client` calls it, it needs to access the block rules which is not available in backend. https://github.com/paritytech/substrate/pull/6301/files#diff-d78994a221a27b69ca7b0d31ff112120R246

So I'm not sure if it's possible to only pass a flag here.

Reply:
`finalization_check` accepts just the block hash. So the client can simply compute the the same closure before calling `commit_operation` and pass the result as a flag. Or, even better, make it a property of `BlockImportOperation`

Reply:
The issue is -- in one `BlockImportOperation` there may be multiple finalization attempts. So that `finalization_check` may be called multiple times with different result.

Reply:
So maybe simply remove the blocks that need not to be finalized from `transaction. finalized_blocks`? The point is that all the data needs to be commited should be in `transaction` already. Introducing additional filters at this stage complicates the interface and makes the logic hard to follow.

Reply:
Now that we directly insert the finalization blacklist to BadBlock, this is not an issue any more!